---
title: "Add annotation to comb-p results generated externally."
author:
  - David Lukacsovich^[University of Miami]
  - Lily Wang^[University of Miami]
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  rmarkdown::html_document:
    highlight: breezedark
    theme: lumen
    toc: true
    number_sections: true
    df_print: paged
    code_download: false
    toc_float:
      collapsed: yes
    toc_depth: 3
editor_options:
  chunk_output_type: inline    
---

# Set Up

## Load Libraries

```{r setup, include=FALSE}
library(readxl)
library(rGREAT)
library(dplyr)
library(GenomicRanges)
library(IRanges)

base_dir <- "~/TBL Dropbox/David Lukacsovich"
analysis_dir <- file.path(base_dir, "AD-Aging-brain-sample-analysis")
mike_dir <- file.path(analysis_dir, "Mike")
code_dir <- file.path(analysis_dir, "code")
reference_dir <- file.path(code_dir, "Reference_Files")
```

# Annotate Comb-p Results

## Load Data

```{r comb_load}
comb_df <- readxl::read_excel(file.path(mike_dir, "cnew.regions-p.bed.xlsx")) %>%
  dplyr::mutate(`#chrom` = as.character(.data$`#chrom`))
comb_df$index <- 1:nrow(comb_df)
```

## GREAT Annotation

```{r comb_great}
# create GRanges object
comb_gr <- GenomicRanges::GRanges(
  seqnames = paste0("chr", comb_df$`#chrom`),
  ranges = IRanges::IRanges(start = comb_df$start, end = comb_df$end)
)

# Submit GREAT job
job <- rGREAT::submitGreatJob(comb_gr, species = "hg19")
regionsToGenes_gr <- rGREAT::getRegionGeneAssociations(job)
regionsToGenes <- as.data.frame(regionsToGenes_gr)
regionsToGenes$cpg <- regionsToGenes_gr@ranges@NAMES

GREAT_annotation <- lapply(
  seq_len(length(regionsToGenes$annotated_genes)),
  function(i) {
    g <- ifelse(regionsToGenes$dist_to_TSS[[i]] > 0,
           paste0(regionsToGenes$annotated_genes[[i]],
                  " (+", regionsToGenes$dist_to_TSS[[i]], ")"),
           paste0(regionsToGenes$annotated_genes[[i]],
                  " (", regionsToGenes$dist_to_TSS[[i]], ")"))
    paste0(g, collapse = ";")
  }
)

great <- dplyr::select(
  regionsToGenes, "seqnames", "start", "end")
great <- data.frame(
  great, GREAT_annotation = unlist(GREAT_annotation)) %>%
  dplyr::mutate(seqnames = as.character(.data$seqnames)) %>%
  dplyr::mutate(seqnames = substring(.data$seqnames, 4, nchar(.data$seqnames)))

comb_df <- comb_df %>%
  dplyr::left_join(great, by = c("#chrom" = "seqnames", "start", "end"))
```

## Illumina Annotation

```{r comb_illumn}
get_array_annotations <- function(array = c("HM450", "EPIC")) {
  array <- match.arg(array)
  
  if (array == "HM450"){
    minfi_object <- utils::data(
      "IlluminaHumanMethylation450kanno.ilmn12.hg19",
      package = "IlluminaHumanMethylation450kanno.ilmn12.hg19"
    )
  } else {
    minfi_object <- utils::data(
      "IlluminaHumanMethylationEPICanno.ilm10b4.hg19",
      package = "IlluminaHumanMethylationEPICanno.ilm10b4.hg19"
    )
  }
  
  anno_df <- minfi_object %>%
    minfi::getAnnotation() %>%
    GenomicRanges::makeGRangesFromDataFrame(
      start.field = "pos", end.field = "pos", keep.extra.columns = TRUE
    ) %>%
    as.data.frame() %>%
    dplyr::select(
      "Name",
      "seqnames",
      "start",
      "end",
      "UCSC_RefGene_Group",
      "UCSC_RefGene_Accession",
      "UCSC_RefGene_Name",
      "Relation_to_Island"
    )
  
  anno_df
}

get_illumina_annotations <- function() {
  anno_epic <- get_array_annotations(array = "EPIC")
  anno_450 <- get_array_annotations(array = "HM450") %>%
    dplyr::filter(!(.data$Name %in% anno_epic$Name))
  
  anno_df <- rbind(anno_epic, anno_450)
  
  anno_df
}

summarise_column <- function(data) {
  data <- data %>%
    stringr::str_split(";") %>%
    unlist() %>%
    unname() %>%
    unique()
  if (length(data) > 0) {
    data <- data[nchar(data) > 0]
  }
  
  data <- paste(data, collapse = ";")
}

anno_df <- get_illumina_annotations()
anno_df$index <- 1:nrow(anno_df)
anno_gr <- GenomicRanges::GRanges(
  seqnames = anno_df$seqnames,
  ranges = IRanges::IRanges(start = anno_df$start, end = anno_df$end)
)
comb_gr <- GenomicRanges::GRanges(
  seqnames = paste0("chr", comb_df$`#chrom`),
  ranges = IRanges::IRanges(
    start = as.integer(comb_df$start), end = as.integer(comb_df$end)
  )
)
overlap_df <- GenomicRanges::findOverlaps(comb_gr, anno_gr) %>%
  as.data.frame()

summ_df <- anno_df %>%
  dplyr::inner_join(overlap_df, by = c("index" = "subjectHits")) %>%
  dplyr::group_by(.data$queryHits) %>%
  dplyr::summarise(
    UCSC_RefGene_Group = summarise_column(.data$UCSC_RefGene_Group),
    UCSC_RefGene_Accession = summarise_column(.data$UCSC_RefGene_Accession),
    UCSC_RefGene_Name = summarise_column(.data$UCSC_RefGene_Name),
    Relation_to_Island = summarise_column(.data$Relation_to_Island)
  )
comb_df <- comb_df %>%
  dplyr::left_join(summ_df, by=c("index" = "queryHits"))
```

## Chrommatin Annotation

```{r comb_chromm}
reference_df <- data.frame(
  abbreviation = c(
    '1_TssA',
    '2_TssAFlnk',
    '3_TxFlnk',
    '4_Tx',
    '5_TxWk',
    '6_EnhG',
    '7_Enh',
    '8_ZNF/Rpts',
    '9_Het',
    '10_TssBiv',
    '11_BivFlnk',
    '12_EnhBiv',
    '13_ReprPC',
    '14_ReprPCWk',
    '15_Quies'
  ),
  state = c(
    "Active TSS",
    "Flanking Active TSS",
    "Transcr. at gene 5' and 3'",
    "Strong transcription",
    "Weak transcription",
    "Genic enhancers",
    "Enhancers",
    "ZNF genes & repeats",
    "Heterochromatin",
    "Bivalent/Poised TSS",
    "Flanking Bivalent TSS/Enh",
    "Bivalent Enhancer",
    "Repressed PolyComb",
    "Weak Repressed PolyComb",
    "Quiescent/Low"
  )
)

chromm_df <- read.table(file.path(reference_dir, "E073_15_coreMarks_mnemonics.bed.gz"), sep = "\t") %>%
  dplyr::rename(seqnames = "V1", start = "V2", end = "V3", abbreviation = "V4") %>%
  dplyr::left_join(reference_df, by = c("abbreviation"))
chromm_df$index <- 1:nrow(chromm_df)

chromm_gr <- GenomicRanges::GRanges(
  seqnames = chromm_df$seqnames,
  ranges = IRanges::IRanges(start = chromm_df$start, end = chromm_df$end)
)

overlap_df <- GenomicRanges::findOverlaps(comb_gr, chromm_gr) %>%
  as.data.frame()

summ_df <- chromm_df %>%
  dplyr::left_join(overlap_df, by = c("index" = "subjectHits")) %>%
  dplyr::group_by(.data$queryHits) %>%
  dplyr::summarise(state = paste(unique(.data$state), collapse = ";"))
comb_df <- comb_df %>%
  dplyr::left_join(summ_df, by=c("index" = "queryHits"))
```

# Probe Directions

## Add Probe Directions in Regions

```{r}
cpg_df <- read.csv(
  file.path(
    analysis_dir,
    "analysis-results",
    "meta_analysis",
    "meta_analysis_single_cpg_bacon.csv"
  )
) %>%
  dplyr::mutate(
    direction = ifelse(
      .data$estimate_bacon == 0,
      ".",
      ifelse(.data$estimate_bacon > 0, "+", "-")
    )
  ) %>%
  dplyr::select(
    "probe",
    "seqnames",
    "start",
    "end",
    "direction"
  )
cpg_df$index <- 1:nrow(cpg_df)

cpg_gr <- GenomicRanges::GRanges(
  seqnames = cpg_df$seqnames,
  ranges = IRanges::IRanges(start = cpg_df$start, end = cpg_df$end)
)
comb_gr <- GenomicRanges::GRanges(
  seqnames = paste0("chr", comb_df$`#chrom`),
  ranges = IRanges::IRanges(
    start = comb_df$start,
    end = comb_df$end
  )
)

overlap_df <- GenomicRanges::findOverlaps(comb_gr, cpg_gr) %>%
  as.data.frame()
direction_df <- overlap_df %>%
  dplyr::left_join(cpg_df, by = c("subjectHits" = "index")) %>%
  dplyr::group_by(.data$queryHits) %>%
  dplyr::summarise(
    combp_direction = paste0(.data$direction, collapse = ""),
    combp_probes = paste0(.data$probe, collapse = ";")
  )

comb_df <- comb_df %>%
  dplyr::filter(.data$index %in% direction_df$queryHits) %>%
  dplyr::left_join(direction_df, by = c("index" = "queryHits")) %>%
  dplyr::select(-"index")
```


# Save

```{r comb_save}
write.csv(
  comb_df,
  file = file.path(mike_dir, "cnew.regions-p.bed.annotated.csv"),
  row.names = FALSE
)
```

# Session Information

<details>
  <summary>**Session Info**</summary>
```{r session}
sessionInfo()
```
</details>
