---
title: "Pathway analysis of CpGs significant by coMethDMR and comb-p"
author:
  - David Lukacsovich^[University of Miami]
  - Lily Wang^[University of Miami]
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  rmarkdown::html_document:
    highlight: breezedark
    theme: lumen
    toc: true
    number_sections: true
    df_print: paged
    code_download: false
    toc_float:
      collapsed: yes
    toc_depth: 3
editor_options:
  chunk_output_type: inline    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Set Up

## Load Libraries

```{r library}
suppressMessages({
  library(DT)
  library(dplyr)
  library(tidyr)
  library(stringr)
  library(GenomicRanges)
  library(IRanges)
  library(missMethyl)
  library(LOLA)
  library(ggplot2)
  library(ggpubr)
})

base_dir <- "~/TBL Dropbox/David Lukacsovich"
analysis_dir <- file.path(base_dir, "AD-Aging-brain-sample-analysis")
lola_dir <- file.path(analysis_dir, "LOLACore")
result_dir <- file.path(analysis_dir, "analysis-results")
save_dir <- file.path(result_dir, "pathway_enrichment")
code_dir <- file.path(analysis_dir, "code")
reference_dir <- file.path(code_dir, "Reference_Files")
```

## Get CpG Positions

```{r load_cpg_position}
get_annotations <- function(array = c("HM450", "EPIC")) {
  array <- match.arg(array)
  
  if (array == "HM450"){
    minfi_object <- utils::data(
      "IlluminaHumanMethylation450kanno.ilmn12.hg19",
      package = "IlluminaHumanMethylation450kanno.ilmn12.hg19"
    )
  } else {
    minfi_object <- utils::data(
      "IlluminaHumanMethylationEPICanno.ilm10b4.hg19",
      package = "IlluminaHumanMethylationEPICanno.ilm10b4.hg19"
    )
  }
  
  anno_df <- minfi_object %>%
    minfi::getAnnotation() %>%
    GenomicRanges::makeGRangesFromDataFrame(
      start.field = "pos", end.field = "pos", keep.extra.columns = TRUE
    ) %>%
    as.data.frame() %>%
    dplyr::select(probe = "Name", "seqnames", "start", "end") %>%
    dplyr::filter(!(.data$seqnames %in% c("chrX", "chrY")))
  
  anno_df
}

cpg_pos_df <- rbind(
  get_annotations("EPIC"),
  get_annotations("HM450")
) %>%
  dplyr::distinct()
```


## Load Target CpGs

```{r load_initialize}
target_cpgs <- list()
```

### Read Probes

```{r load_probes}
cpg_df <- read.csv(file = file.path(
  analysis_dir,
  "analysis-results",
  "meta_analysis",
  "meta_analysis_single_cpg_bacon.csv"
  )
)

dmr_sig_df <- read.csv(
  file = file.path(
    analysis_dir,
    "analysis-results",
    "meta_analysis",
    "meta_analysis_summary.csv"
  )
)

ranges_450_ls <- readRDS(file.path(reference_dir, "HM450_Ranges.RDS")) %>%
  unlist() %>%
  unname() %>%
  unique() %>%
  intersect(cpg_df$probe)
ranges_epic_ls <- readRDS(file.path(reference_dir, "EPIC_Ranges.RDS")) %>%
  unlist() %>%
  unname() %>%
  unique() %>%
  intersect(cpg_df$probe)

cpg_df <- cpg_df %>%
  dplyr::filter(!is.na(.data$BDR_estimate)) %>%
  dplyr::filter(!is.na(.data$ROSMAP_estimate))
```

### Get DMR Probes

```{r load_dmr_probes}
target_cpgs[["DMR_bg"]] <- c(
  ranges_450_ls %>% unlist() %>% unname() %>% unique(),
  ranges_epic_ls %>% unlist() %>% unname() %>% unique()
) %>%
  unique()

target_cpgs[["DMR_fg"]] <- targets <- dmr_sig_df %>%
  dplyr::pull("combp_probes") %>%
  stringr::str_split(";") %>%
  unlist() %>%
  unique()

target_cpgs[["DMR_hyper"]] <- dmr_sig_df %>%
  dplyr::filter(.data$direction_bacon == "++") %>%
  dplyr::pull("combp_probes") %>%
  stringr::str_split(";") %>%
  unlist() %>%
  unique()

target_cpgs[["DMR_hypo"]] <- dmr_sig_df %>%
  dplyr::filter(.data$direction_bacon == "--") %>%
  dplyr::pull("combp_probes") %>%
  stringr::str_split(";") %>%
  unlist() %>%
  unique()
```

### Get Single CpG Probes

```{r load_cpg_probes}
target_cpgs[["single_bg"]] <- cpg_df$probe
target_cpgs[["single_fg"]] <- cpg_df %>%
  dplyr::filter(.data$fdr_bacon_inclusive < 0.05) %>%
  dplyr::filter(.data$direction_bacon %in% c("--", "++")) %>%
  dplyr::pull("probe")
target_cpgs[["single_hyper"]] <- cpg_df %>%
  dplyr::filter(.data$fdr_bacon_inclusive < 0.05) %>%
  dplyr::filter(.data$direction_bacon == "++") %>%
  dplyr::pull("probe")
target_cpgs[["single_hypo"]] <- cpg_df %>%
  dplyr::filter(.data$fdr_bacon_inclusive < 0.05) %>%
  dplyr::filter(.data$direction_bacon == "--") %>%
  dplyr::pull("probe")
```

### Get Full Probes Lists

```{r load_all_probes}
target_cpgs[["bg"]] <- unique(c(target_cpgs$single_bg, target_cpgs$DMR_bg))
target_cpgs[["fg"]] <- unique(c(target_cpgs$single_fg, target_cpgs$DMR_fg))
target_cpgs[["hyper"]] <- unique(c(target_cpgs$single_hyper, target_cpgs$DMR_hyper))
target_cpgs[["hypo"]] <- unique(c(target_cpgs$single_hypo, target_cpgs$DMR_hypo))

lengths(target_cpgs)
```


# Get Ontologies

## Get GO Ontologies

```{r ont_go, eval=FALSE}
all_go <- missMethyl::gometh(
  sig.cpg = target_cpgs$fg,
  all.cpg = target_cpgs$bg,
  collection = "GO",
  fract.counts = TRUE,
  sig.genes = TRUE
) %>%
  tibble::rownames_to_column(var = "GO") %>%
  dplyr::arrange(.data$P.DE) %>%
  dplyr::filter(.data$N <= 200) %>%
  dplyr::filter(.data$N >= 5) %>%
  dplyr::select(
    Gene_Set = "GO",
    Ontology = "ONTOLOGY",
    Term = "TERM",
    Gene_Count = "N",
    "DE",
    pValue = "P.DE",
    fdr = "FDR",
    de_genes = "SigGenesInSet"
  ) %>%
  dplyr::mutate(
    fdr = p.adjust(.data$pValue,"fdr")
  )

bp_go <- all_go %>%
  dplyr::filter(.data$Ontology == "BP") %>%
  dplyr::mutate(FDR = p.adjust(.data$pValue,"fdr"))
```

## Get KEGG Ontologies

```{r ont_kegg, eval=FALSE}
all_kegg <- missMethyl::gometh(
  sig.cpg = target_cpgs$fg,
  all.cpg = target_cpgs$bg,
  collection = "KEGG",
  fract.counts = TRUE,
  sig.genes = TRUE
) %>%
  tibble::rownames_to_column(var = "KEGG") %>%
  dplyr::arrange(.data$P.DE) %>%
  dplyr::filter(.data$N <= 200) %>%
  dplyr::filter(.data$N >= 5) %>%
  dplyr::select(
    Gene_Set = "KEGG",
    Term = "Description",
    Gene_Count = "N",
    "DE",
    pValue = "P.DE",
    fdr = "FDR",
    de_genes = "SigGenesInSet"
  ) %>%
  dplyr::mutate(
    fdr = p.adjust(.data$pValue,"fdr")
  )
```

## Save Results

```{r ont_save, eval=FALSE}
write.csv(
  all_go,
  file = file.path(save_dir, "sig_pathways_GO.csv"),
  row.names = FALSE
)
write.csv(
  bp_go,
  file = file.path(save_dir, "sig_pathways_BP.csv"),
  row.names = FALSE
)
write.csv(
  all_kegg,
  file = file.path(save_dir, "sig_pathways_KEGG.csv"),
  row.names = FALSE
)
```

# Region Enrichment

## Update Annotations

### Get Illumina Annotations

```{r annotate_illum}
# Island Annotations
minfi_object <- utils::data(
  "IlluminaHumanMethylationEPICanno.ilm10b4.hg19",
  package = "IlluminaHumanMethylationEPICanno.ilm10b4.hg19"
)
anno_EPIC <- minfi::getAnnotation(minfi_object) %>%
  as.data.frame() %>%
  dplyr::select(
    probe = "Name",
    Island = "Relation_to_Island",
    Group = "UCSC_RefGene_Group"
  ) %>%
  dplyr::mutate(
    Group = stringr::str_split(.data$Group, ";")) %>%
  tidyr::unnest("Group") %>%
  dplyr::distinct()

minfi_object <- utils::data(
  "IlluminaHumanMethylation450kanno.ilmn12.hg19",
  package = "IlluminaHumanMethylation450kanno.ilmn12.hg19"
)
anno_450 <- minfi::getAnnotation(minfi_object) %>%
  as.data.frame() %>%
  dplyr::select(
    probe = "Name",
    Island = "Relation_to_Island",
    Group = "UCSC_RefGene_Group"
  ) %>%
  dplyr::mutate(
    Group = stringr::str_split(.data$Group, ";")) %>%
  tidyr::unnest("Group") %>%
  dplyr::distinct() %>%
  dplyr::filter(!(.data$probe %in% anno_EPIC$probe))

anno_df <- rbind(
  anno_EPIC,
  anno_450
) %>%
  dplyr::mutate(
    Island = gsub("N_|S_", "", .data$Island)
  ) %>%
  dplyr::mutate(
    Group = ifelse(nchar(.data$Group) == 0, "Intergenic", .data$Group)
  )
```

### Get Chrommatin Annotations

```{r annotate_chromm}
reference_df <- data.frame(
  abbreviation = c(
    '1_TssA',
    '2_TssAFlnk',
    '3_TxFlnk',
    '4_Tx',
    '5_TxWk',
    '6_EnhG',
    '7_Enh',
    '8_ZNF/Rpts',
    '9_Het',
    '10_TssBiv',
    '11_BivFlnk',
    '12_EnhBiv',
    '13_ReprPC',
    '14_ReprPCWk',
    '15_Quies'
  ),
  state = c(
    "Active TSS",
    "Flanking Active TSS",
    "Transcr. at gene 5' and 3'",
    "Strong transcription",
    "Weak transcription",
    "Genic enhancers",
    "Enhancers",
    "ZNF genes & repeats",
    "Heterochromatin",
    "Bivalent/Poised TSS",
    "Flanking Bivalent TSS/Enh",
    "Bivalent Enhancer",
    "Repressed PolyComb",
    "Weak Repressed PolyComb",
    "Quiescent/Low"
  )
)

chromm_df <- read.table(file.path(reference_dir, "E073_15_coreMarks_mnemonics.bed.gz"), sep = "\t") %>%
  dplyr::rename(seqnames = "V1", start = "V2", end = "V3", abbreviation = "V4") %>%
  dplyr::left_join(reference_df, by = c("abbreviation"))
```

### Add Abbreviations

```{r add_annotate}
cpg_pos_df <- cpg_pos_df %>%
  dplyr::left_join(anno_df, by = "probe")
cpg_pos_df$index <- 1:nrow(cpg_pos_df)
state_df <- chromm_df %>%
  dplyr::select(State = "state") %>%
  dplyr::mutate(index = 1:nrow(chromm_df))

state_gr <- GenomicRanges::GRanges(
  seqnames = chromm_df$seqnames,
  ranges = IRanges::IRanges(start = chromm_df$start, end = chromm_df$end)
)
cpg_gr <- GenomicRanges::GRanges(
  seqnames = cpg_pos_df$seqnames,
  ranges = IRanges::IRanges(start = cpg_pos_df$start, end = cpg_pos_df$end)
)

overlap_df <- GenomicRanges::findOverlaps(cpg_gr, state_gr) %>%
  as.data.frame()

cpg_pos_df <- cpg_pos_df %>%
  dplyr::left_join(overlap_df, by = c("index" = "queryHits")) %>%
  dplyr::left_join(state_df, by = c("subjectHits" = "index")) %>%
  dplyr::select(-"index", -"subjectHits")
```

## Define Annotation Function

```{r annotate_functions}
get_cpg_region_summary <-  function(fg_probes, bg_probes, anno_df, label_col)
{
  anno_df <- anno_df %>%
    dplyr::select("probe", label = !!as.name(label_col)) %>%
    dplyr::distinct()
  
  fg_counts <- anno_df %>%
    dplyr::filter(.data$probe %in% fg_probes) %>%
    dplyr::select("label") %>%
    plyr::count() %>%
    as.data.frame.matrix() %>%
    dplyr::rename(foreground = "freq")
  
  bg_counts <- anno_df %>%
    dplyr::filter(.data$probe %in% bg_probes) %>%
    dplyr::select("label") %>%
    plyr::count() %>%
    as.data.frame.matrix() %>%
    dplyr::rename(background = "freq")
  
  res_df <- fg_counts %>%
    dplyr::full_join(bg_counts, by = "label") %>%
    tidyr::replace_na(list(foreground = 0, background = 0)) %>%
    dplyr::mutate(
      fg_total = length(unique(fg_probes)),
      bg_total = length(unique(bg_probes))
    )
  
  res_df
}

get_cpg_fisher_stats <- function(res_df) {
  fish_df <- plyr::ldply(
    .data = 1:nrow(res_df), .fun = function(idx) {
      a <- res_df[idx, "foreground"]
      # b <- res_df$[-idx, "foreground] %>% sum()
      b <- res_df[idx, "fg_total"] - a
      c <- res_df[idx, "background"] - a 
      # d <- (res_df[-idx,3] %>% sum()) - b
      d <- res_df[idx, "bg_total"] - res_df[idx, "background"] - b
      
      mat <- matrix(
        c(a,c,b,d),
        nrow = 2, 
        dimnames = list(type = c("Yes","No"), dir = c("Yes","No"))
      )
      
      ft <- stats::fisher.test(mat, alternative = "two.sided")
      df <- data.frame(
        foreground_count = a,
        background_count = c,
        foreground_frequency = 100 * a / (a + b),
        background_frequency = 100 * (a + c) / (a + b + c + d),
        p_value = ft$p.value,
        odds_ratio = ft$estimate,
        a = a,
        b = b,
        c = c,
        d = d
      )
      return(df)
      
    }, .id = NULL)
  
  fish_df
}

get_cpg_region_enrichment <- function(fg_probes, bg_probes, anno_df, label_col)
{
  res_df <- get_cpg_region_summary(fg_probes, bg_probes, anno_df, label_col)
  fish_df <- get_cpg_fisher_stats(res_df)
  fish_df[,"features"] = res_df$label
  fish_df <- fish_df %>%
    dplyr::select(
      "features",
      "foreground_count",
      "background_count",
      "foreground_frequency",
      "background_frequency",
      "p_value",
      "odds_ratio",
      "a",
      "b",
      "c",
      "d"
    )
  
  fish_df
}

get_cpg_region_comparisons <- function(
    target_cpgs, category, anno_df, label_col, label
) {
  all_df <- get_cpg_region_enrichment(
    fg_probes = target_cpgs[[paste0(category, "fg")]],
    bg_probes = target_cpgs[[paste0(category, "bg")]],
    anno_df = anno_df,
    label_col = label_col
  ) %>%
    dplyr::mutate(
      analysis = sprintf(label, "All"), .before = "features"
    )
  
  hyper_df <- get_cpg_region_enrichment(
    target_cpgs[[paste0(category, "hyper")]],
    target_cpgs[[paste0(category, "bg")]],
    anno_df,
    label_col
  ) %>%
    dplyr::mutate(
      analysis = sprintf(label, "Hyper"), .before = "features"
    )
  
  hypo_df <- get_cpg_region_enrichment(
    target_cpgs[[paste0(category, "hypo")]],
    target_cpgs[[paste0(category, "bg")]],
    anno_df,
    label_col
  ) %>%
    dplyr::mutate(
      analysis = sprintf(label, "Hypo"), .before = "features"
    )
  
  rbind(
    all_df,
    hyper_df,
    hypo_df
  )
}

get_cpg_region_plots <- function(region_df, xlab, angle = 0) {
  plot_df <- region_df %>%
    dplyr::select(
      "analysis",
      "features",
      "foreground_frequency",
      "background_frequency"
    ) %>%
    tidyr::pivot_longer(
      cols = c("foreground_frequency", "background_frequency"),
      values_to = "frequency",
      names_to = "probe"
    )
  
  plot <- ggpubr::ggbarplot(
    plot_df,
    x = "features",
    y = "frequency",
    fill = "probe", 
    color = "white",
    ylab = "Frequency (% Counts)",
    xlab = xlab,
    position = ggplot2::position_dodge(0.8)
  ) +
    ggplot2::theme(legend.title = ggplot2::element_blank()) + 
    ggplot2::scale_fill_manual(
      name = "", 
      labels = c("All Probes", "Significant Probes"),
      values = c(
        "background_frequency" = "#1F77B4", 
        "foreground_frequency" = "#FF7F0D"
      )
    ) +
    ggpubr::rotate_x_text(angle = angle)
  
  ggpubr::facet(plot, facet.by = "analysis", ncol = 1)
}
```

## Evaluate

### Evaluate DMRs

#### Get Analysis Tables

```{r enrich_dmr_get}
island_df <- get_cpg_region_comparisons(
  target_cpgs,
  "DMR_",
  cpg_pos_df,
  "Island",
  label = "Relation to Island - %s sig. combpDMR vs clustered probes"
)

group_df <- get_cpg_region_comparisons(
  target_cpgs,
  "DMR_",
  cpg_pos_df,
  "Group",
  label = "UCSC RefGene Group hierarchy - %s sig. combpDMR vs clustered probes"
)

chromm_df <- get_cpg_region_comparisons(
  target_cpgs,
  "DMR_",
  cpg_pos_df,
  "State",
  label = "Chromatin State - %s sig. combpDMR vs clustered probes"
)
```

#### Plot Tables

```{r enrich_dmr_plot_island, fig.height = 8}
get_cpg_region_plots(island_df, "Relation to island")
```

```{r enrich_dmr_show_island}
island_df %>%
  dplyr::mutate(
    foreground_frequency = round(.data$foreground_frequency, digits = 2),
    background_frequency = round(.data$background_frequency, digits = 2),
    odds_ratio = round(.data$odds_ratio, digits = 2),
    p_value = signif(.data$p_value, digits = 3)
  ) %>%
  DT::datatable(
    style = "bootstrap",
    extensions = 'Buttons',
    options = list(
      scrollX = TRUE, 
      dom = 'Bfrtip',
      buttons = I('colvis'),
      keys = TRUE, 
      pageLength = 10
    ), 
    rownames = FALSE
  )
```

```{r enrich_dmr_plot_group, fig.height = 8}
get_cpg_region_plots(group_df, "UCSC RefGene Group")
```

```{r enrich_dmr_show_group}
group_df %>%
  dplyr::mutate(
    foreground_frequency = round(.data$foreground_frequency, digits = 2),
    background_frequency = round(.data$background_frequency, digits = 2),
    odds_ratio = round(.data$odds_ratio, digits = 2),
    p_value = signif(.data$p_value, digits = 3)
  ) %>%
  DT::datatable(
    style = "bootstrap",
    extensions = 'Buttons',
    options = list(
      scrollX = TRUE, 
      dom = 'Bfrtip',
      buttons = I('colvis'),
      keys = TRUE, 
      pageLength = 10
    ), 
    rownames = FALSE
  )
```

```{r enrich_dmr_plot_state, fig.height = 8}
get_cpg_region_plots(chromm_df, "Chromatin State", angle = 90)
```

```{r enrich_dmr_show_state}
chromm_df %>%
  dplyr::mutate(
    foreground_frequency = round(.data$foreground_frequency, digits = 2),
    background_frequency = round(.data$background_frequency, digits = 2),
    odds_ratio = round(.data$odds_ratio, digits = 2),
    p_value = signif(.data$p_value, digits = 3)
  ) %>%
  DT::datatable(
    style = "bootstrap",
    extensions = 'Buttons',
    options = list(
      scrollX = TRUE, 
      dom = 'Bfrtip',
      buttons = I('colvis'),
      keys = TRUE, 
      pageLength = 10
    ), 
    rownames = FALSE
  )
```

#### Save Tables

```{r enrich_dmr_save}
illum_df <- rbind(island_df, group_df)

write.csv(
  illum_df,
  file = file.path(save_dir, "sig_enrichment_dmr_illumina.csv"),
  row.names = FALSE
)

write.csv(
  chromm_df,
  file = file.path(save_dir, "sig_enrichment_dmr_chrommatin.csv"),
  row.names = FALSE
)
```

### Evaluate single CpGs

#### Get Analysis Tables

```{r enrich_cpg_get}
island_df <- get_cpg_region_comparisons(
  target_cpgs,
  "single_",
  cpg_pos_df,
  "Island",
  label = "Relation to Island - %s sig. probes vs all probes"
)

group_df <- get_cpg_region_comparisons(
  target_cpgs,
  "single_",
  cpg_pos_df,
  "Group",
  label = "UCSC RefGene Group hierarchy - %s sig. probes vs all probes"
)

chromm_df <- get_cpg_region_comparisons(
  target_cpgs,
  "single_",
  cpg_pos_df,
  "State",
  label = "Chromatin State - %s sig. probes vs all probes"
)
```

#### Plot Tables

```{r enrich_cpg_plot_island, fig.height = 8}
get_cpg_region_plots(island_df, "Relation to island")
```

```{r enrich_cpg_show_island}
island_df %>%
  dplyr::mutate(
    foreground_frequency = round(.data$foreground_frequency, digits = 2),
    background_frequency = round(.data$background_frequency, digits = 2),
    odds_ratio = round(.data$odds_ratio, digits = 2),
    p_value = signif(.data$p_value, digits = 3)
  ) %>%
  DT::datatable(
    style = "bootstrap",
    extensions = 'Buttons',
    options = list(
      scrollX = TRUE, 
      dom = 'Bfrtip',
      buttons = I('colvis'),
      keys = TRUE, 
      pageLength = 10
    ), 
    rownames = FALSE
  )
```

```{r enrich_cpg_plot_group, fig.height = 8}
get_cpg_region_plots(group_df, "UCSC RefGene Group")
```

```{r enrich_cpg_show_group}
group_df %>%
  dplyr::mutate(
    foreground_frequency = round(.data$foreground_frequency, digits = 2),
    background_frequency = round(.data$background_frequency, digits = 2),
    odds_ratio = round(.data$odds_ratio, digits = 2),
    p_value = signif(.data$p_value, digits = 3)
  ) %>%
  DT::datatable(
    style = "bootstrap",
    extensions = 'Buttons',
    options = list(
      scrollX = TRUE, 
      dom = 'Bfrtip',
      buttons = I('colvis'),
      keys = TRUE, 
      pageLength = 10
    ), 
    rownames = FALSE
  )
```

```{r enrich_cpg_plot_state, fig.height = 8}
get_cpg_region_plots(chromm_df, "Chromatin State", angle = 90)
```

```{r enrich_cpg_show_state}
chromm_df %>%
  dplyr::mutate(
    foreground_frequency = round(.data$foreground_frequency, digits = 2),
    background_frequency = round(.data$background_frequency, digits = 2),
    odds_ratio = round(.data$odds_ratio, digits = 2),
    p_value = signif(.data$p_value, digits = 3)
  ) %>%
  DT::datatable(
    style = "bootstrap",
    extensions = 'Buttons',
    options = list(
      scrollX = TRUE, 
      dom = 'Bfrtip',
      buttons = I('colvis'),
      keys = TRUE, 
      pageLength = 10
    ), 
    rownames = FALSE
  )
```

#### Save Tables

```{r enrich_cpg_save}
illum_df <- rbind(island_df, group_df)

write.csv(
  illum_df,
  file = file.path(save_dir, "sig_enrichment_cpg_illumina.csv"),
  row.names = FALSE
)

write.csv(
  chromm_df,
  file = file.path(save_dir, "sig_enrichment_cpg_chrommatin.csv"),
  row.names = FALSE
)
```

### Evaluate DMRs and single CpGs

#### Get Analysis Tables

```{r enrich_both_get}
island_df <- get_cpg_region_comparisons(
  target_cpgs,
  "",
  cpg_pos_df,
  "Island",
  label = "Relation to Island - %s sig. probes + combpDMR vs all probes"
)

group_df <- get_cpg_region_comparisons(
  target_cpgs,
  "",
  cpg_pos_df,
  "Group",
  label = "UCSC RefGene Group hierarchy - %s sig. probes + combpDMR vs all probes"
)

chromm_df <- get_cpg_region_comparisons(
  target_cpgs,
  "",
  cpg_pos_df,
  "State",
  label = "Chromatin State - %s sig. probes + combpDMR vs all probes"
)
```

#### Plot Tables

```{r enrich_both_plot_island, fig.height = 8}
get_cpg_region_plots(island_df, "Relation to island")
```

```{r enrich_both_show_island}
island_df %>%
  dplyr::mutate(
    foreground_frequency = round(.data$foreground_frequency, digits = 2),
    background_frequency = round(.data$background_frequency, digits = 2),
    odds_ratio = round(.data$odds_ratio, digits = 2),
    p_value = signif(.data$p_value, digits = 3)
  ) %>%
  DT::datatable(
    style = "bootstrap",
    extensions = 'Buttons',
    options = list(
      scrollX = TRUE, 
      dom = 'Bfrtip',
      buttons = I('colvis'),
      keys = TRUE, 
      pageLength = 10
    ), 
    rownames = FALSE
  )
```

```{r enrich_both_plot_group, fig.height = 8}
get_cpg_region_plots(group_df, "UCSC RefGene Group")
```

```{r enrich_both_show_group}
group_df %>%
  dplyr::mutate(
    foreground_frequency = round(.data$foreground_frequency, digits = 2),
    background_frequency = round(.data$background_frequency, digits = 2),
    odds_ratio = round(.data$odds_ratio, digits = 2),
    p_value = signif(.data$p_value, digits = 3)
  ) %>%
  DT::datatable(
    style = "bootstrap",
    extensions = 'Buttons',
    options = list(
      scrollX = TRUE, 
      dom = 'Bfrtip',
      buttons = I('colvis'),
      keys = TRUE, 
      pageLength = 10
    ), 
    rownames = FALSE
  )
```

```{r enrich_both_plot_state, fig.height = 8}
get_cpg_region_plots(chromm_df, "Chromatin State", angle = 90)
```

```{r enrich_both_show_state}
chromm_df %>%
  dplyr::mutate(
    foreground_frequency = round(.data$foreground_frequency, digits = 2),
    background_frequency = round(.data$background_frequency, digits = 2),
    odds_ratio = round(.data$odds_ratio, digits = 2),
    p_value = signif(.data$p_value, digits = 3)
  ) %>%
  DT::datatable(
    style = "bootstrap",
    extensions = 'Buttons',
    options = list(
      scrollX = TRUE, 
      dom = 'Bfrtip',
      buttons = I('colvis'),
      keys = TRUE, 
      pageLength = 10
    ), 
    rownames = FALSE
  )
```

#### Save Tables

```{r enrich_both_save}
illum_df <- rbind(island_df, group_df)

write.csv(
  illum_df,
  file = file.path(save_dir, "sig_enrichment_both_illumina.csv"),
  row.names = FALSE
)

write.csv(
  chromm_df,
  file = file.path(save_dir, "sig_enrichment_both_chrommatin.csv"),
  row.names = FALSE
)
```


# LOLA

## Load DB

```{r lola_db, message = FALSE, echo = FALSE}
regionDB_hg19 <- LOLA::loadRegionDB(file.path(lola_dir, "hg19"))
```

## Define Functions

```{r lola_functions}
get_lola_region_enrichment <- function(fg_probes, bg_probes, anno_df) {
  fg_gr <- cpg_pos_df %>%
    dplyr::filter(.data$probe %in% fg_probes) %>%
    GenomicRanges::makeGRangesFromDataFrame(
      seqnames.field = "seqnames",
      start.field = "start",
      end.field = "end"
    )
  bg_gr <- cpg_pos_df %>%
    dplyr::filter(.data$probe %in% bg_probes) %>%
    GenomicRanges::makeGRangesFromDataFrame(
      seqnames.field = "seqnames",
      start.field = "start",
      end.field = "end"
    )
  
  locResults <- LOLA::runLOLA(
    userSets = fg_gr,
    userUniverse = bg_gr,
    regionDB = regionDB_hg19,
    cores = 8
  ) %>%
    dplyr::mutate(
      pValue = 10^(-.data$pValueLog)
    )
  
  locResults
}

get_lola_region_comparisons <- function(
    target_cpgs, category, anno_df
) {
  all_df <- get_lola_region_enrichment(
    fg_probes = target_cpgs[[paste0(category, "fg")]],
    bg_probes = target_cpgs[[paste0(category, "bg")]],
    anno_df = anno_df
  ) %>%
    dplyr::select(
    Antibody = "antibody",
    CellType = "cellType",
    FileName = "filename",
    "collection",
    OR = "oddsRatio",
    "pValue",
    FDR = "qValue"
  )
  
  hyper_df <- get_lola_region_enrichment(
    target_cpgs[[paste0(category, "hyper")]],
    target_cpgs[[paste0(category, "bg")]],
    anno_df
  ) %>%
    dplyr::select(
    Antibody = "antibody",
    CellType = "cellType",
    FileName = "filename",
    "collection",
    hyper_OR = "oddsRatio",
    hyper_pValue = "pValue",
    hyper_FDR = "qValue"
  )
  
  hypo_df <- get_lola_region_enrichment(
    target_cpgs[[paste0(category, "hypo")]],
    target_cpgs[[paste0(category, "bg")]],
    anno_df
  ) %>%
    dplyr::select(
    Antibody = "antibody",
    CellType = "cellType",
    FileName = "filename",
    "collection",
    hypo_OR = "oddsRatio",
    hypo_pValue = "pValue",
    hypo_FDR = "qValue"
  )
  
  all_df %>%
    dplyr::full_join(hyper_df) %>%
    dplyr::full_join(hypo_df) %>%
    dplyr::filter(
      min(.data$FDR, .data$hypo_FDR, .data$hyper_FDR, na.rm = TRUE) < 0.05
    ) %>%
    dplyr::arrange(.data$pValue)
}
```

## Evaluate

```{r lola_eval, eval=FALSE}
dmr_lola <- get_lola_region_comparisons(target_cpgs, "DMR_", cpg_pos_df)
cpg_lola <- get_lola_region_comparisons(target_cpgs, "single_", cpg_pos_df)
both_lola <- get_lola_region_comparisons(target_cpgs, "", cpg_pos_df)

write.csv(
  dmr_lola,
  file = file.path(save_dir, "lola_enrichment_dmr.csv"),
  row.names = FALSE
)

write.csv(
  cpg_lola,
  file = file.path(save_dir, "lola_enrichment_cpg.csv"),
  row.names = FALSE
)

write.csv(
  both_lola,
  file = file.path(save_dir, "lola_enrichment_both.csv"),
  row.names = FALSE
)
```

# Session Information

<details>
  <summary>**Session Info**</summary>
```{r session}
sessionInfo()
```
