---
title: "Get association between DNAm residuals and RNA residuals in ROSMAP data"
subtitle: "We correlate probes and genes on opposite strands"
author:
  - David Lukacsovich^[University of Miami]
  - Lily Wang^[University of Miami]
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  rmarkdown::html_document:
    highlight: breezedark
    theme: lumen
    toc: true
    number_sections: true
    df_print: paged
    code_download: false
    toc_float:
      collapsed: yes
    toc_depth: 3
editor_options:
  chunk_output_type: inline    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Set Up

## Load Libraries

```{r library}
suppressMessages({
  library(dplyr)
  library(tidyr)
  library(tibble)
  library(GenomicRanges)
  library(MethReg)
})

base_dir <- "~/TBL Dropbox/David Lukacsovich"
analysis_dir <- file.path(base_dir, "AD-Aging-brain-sample-analysis")
result_dir <- file.path(analysis_dir, "analysis-results")
code_dir <- file.path(analysis_dir, "code")
dataset_dir <- file.path(code_dir, "DATASETS")
rosmap_dir <- file.path(dataset_dir, "Rosmap")
reference_dir <- file.path(code_dir, "Reference_Files")
rna_dir <- file.path(base_dir, "coMethDMR_metaAnalysis/DNAm_RNA/data")
```

# define functions

```{r functions}
# Match genes to DMR and probes
match_features <- function(dnam_se, dmr_se, rna_se) {
## Get Gene to Region Match

  dmr_df <- dmr_se %>%
    SummarizedExperiment::rowData() %>%
    data.frame()
  dmr_gr <- dmr_df %>%
    GenomicRanges::makeGRangesFromDataFrame()
  
  dnam_df <- dnam_se %>%
    SummarizedExperiment::rowData() %>%
    data.frame() %>%
    dplyr::mutate(start = .data$pos, end = .data$pos, strand = .data$strand_rev)
  dnam_gr <- dnam_df %>%
    GenomicRanges::makeGRangesFromDataFrame()

  dmr_genes <- MethReg::get_region_target_gene(
    regions.gr = dmr_gr,
    genome = "hg19",
    method = "window",
    window.size = 500 * 10 ^ 3,
    rm.promoter.regions.from.distal.linking = FALSE
  ) %>%
    dplyr::filter(.data$target %in% row.names(rna_se)) %>%
    dplyr::mutate(reference = .data$regionID)
  
  dnam_genes <- MethReg::get_region_target_gene(
    regions.gr = dnam_gr,
    genome = "hg19",
    method = "window",
    window.size = 500 * 10 ^ 3,
    rm.promoter.regions.from.distal.linking = FALSE
  ) %>%
    dplyr::filter(.data$target %in% row.names(rna_se)) %>%
    tidyr::separate(
      .data$regionID,
      into = c("chr", "start", "end"),
      sep = ":|-",
      remove = TRUE
    ) %>%
    dplyr::mutate(
      start = as.integer(.data$start),
      end = as.integer(.data$end)
    ) %>%
    dplyr::left_join(
      dnam_df %>% dplyr::select("probe", "chr", "start", "end")
    ) %>%
    dplyr::select(
      reference = "probe", "target", "target_symbol", "distance_region_target_tss"
    )
  
  list(dmr_genes = dmr_genes, dnam_genes = dnam_genes)
}

# Calculate assocation statistics
calculate_association <- function(dnam_se, rna_se, genes_df, adjust_braak = TRUE) {
  dnam_data <- SummarizedExperiment::assay(dnam_se, 1)
  rna_data <- SummarizedExperiment::assay(rna_se, 1)
  
  pheno_df <- dnam_se %>%
    SummarizedExperiment::colData() %>%
    as.data.frame()
  
  resid_res <- plyr::adply(
    genes_df,
    .margins = 1,
    .fun = function(row) {
      tryCatch({
        target_rna <- rna_data[row.names(rna_data) == row$target, , drop = FALSE]
        target_dnam <- dnam_data[row.names(dnam_data) == row$reference, , drop = FALSE]
        
        df <- data.frame(
          resid_rna = as.numeric(target_rna),
          resid_dnam = as.numeric(target_dnam),
          braaksc = as.integer(as.numeric(pheno_df$braaksc))
        )
        
        # fit linear model:
        if (adjust_braak) {
          result_fit <- stats::lm(resid_rna ~ resid_dnam + braaksc, data = df)
        } else {
          result_fit <- stats::lm(resid_rna ~ resid_dnam, data = df)
        }
  
        result_pval <- summary(result_fit)$coefficients["resid_dnam", "Pr(>|t|)"]
        result_estimate <- summary(result_fit)$coefficients["resid_dnam", "Estimate"]
        
        data.frame(
          estimate = result_estimate,
          pvalue = result_pval
        )
      }, error = function(e) {
        # print(row)
        return()
      })
    },
    .id = NULL,
    .progress = "time",
    .parallel = TRUE,
    .inform = TRUE
  ) %>%
  dplyr::mutate(fdr = p.adjust(.data$pvalue, method = "fdr"))
}

# annotate dmr results
annotate_dmr <- function(assoc_dmr_df, dmr_ref, rna_se) {
  rna_df <- rna_se %>%
    SummarizedExperiment::rowData() %>%
    data.frame()
  
  assoc_dmr_df <- assoc_dmr_df %>%
    dplyr::select(
      combp_DMR = "reference",
      ens = "target",
      "estimate",
      "pvalue",
      "fdr"
    )
  
  assoc_df <- assoc_dmr_df %>%
    dplyr::left_join(rna_df, by = "ens") %>%
    dplyr::left_join(dmr_ref, by = "combp_DMR") %>%
    dplyr::select(
      "DMR",
      "combp_DMR",
      "ens",
      "gene",
      "estimate",
      "pvalue",
      "fdr",
      "count_total",
      "count_norm",
      "count_ad",
      "GREAT_annotation",
      "UCSC_RefGene_Group",
      "UCSC_RefGene_Name",
      "Relation_to_Island",
      "state"
    )
  
  assoc_df
}

annotate_dnam <- function(assoc_dnam_df, dnam_ref, rna_se) {
  rna_df <- rna_se %>%
    SummarizedExperiment::rowData() %>%
    data.frame()
  
  assoc_dnam_df <- assoc_dnam_df %>%
    dplyr::select(
      probe = "reference",
      ens = "target",
      "estimate",
      "pvalue",
      "fdr"
    )
  
  assoc_df <- assoc_dnam_df %>%
    dplyr::left_join(rna_df, by = "ens") %>%
    dplyr::left_join(dnam_ref, by = "probe") %>%
    dplyr::select(
      CpG = "probe",
      "ens",
      "gene",
      "estimate",
      "pvalue",
      "fdr",
      Chr = "seqnames",
      position = "start",
      "count_total",
      "count_norm",
      "count_ad",
      "GREAT_annotation",
      "UCSC_RefGene_Group",
      "UCSC_RefGene_Name",
      "Relation_to_Island",
      "state"
    )
  
  assoc_df
}
```

# Load References

We load analysis results of CpGs and DMRs, because they have annotation information

```{r load_reference}
dnam_ref <- read.csv(file = file.path(
  analysis_dir,
  "analysis-results",
  "meta_analysis",
  "meta_analysis_single_cpg_bacon.csv"
  )
) %>%
  dplyr::filter(.data$fdr_bacon_inclusive < 0.05) %>%
  dplyr::filter(.data$direction_bacon %in% c("--", "++")) %>%
  dplyr::select(
    "probe",
    "seqnames",
    "start",
    "GREAT_annotation",
    "UCSC_RefGene_Group",
    "UCSC_RefGene_Name",
    "Relation_to_Island",
    "state"
  )

dmr_ref <- read.csv(
  file = file.path(
    analysis_dir,
    "analysis-results",
    "meta_analysis",
    "meta_analysis_summary.csv"
  )
) %>%
  as.data.frame() %>%
  dplyr::select(
    "DMR",
    "combp_DMR",
    "GREAT_annotation",
    "UCSC_RefGene_Group",
    "UCSC_RefGene_Name",
    "Relation_to_Island",
    "state"
  )
```

# Analyze Data

## All Data

### load data

```{r load_all}
dnam_se <- readRDS(file = file.path(rosmap_dir, "rna_dnam", "dnam_data.RDS"))
dmr_se <- readRDS(file = file.path(rosmap_dir, "rna_dnam", "dmr_data.RDS"))
rna_se <- readRDS(file = file.path(rosmap_dir, "rna_dnam", "rnaseq_data.RDS"))
```

### get associations

```{r calc_assocation_all, eval = TRUE}
# match features
res_match <- match_features(dnam_se, dmr_se, rna_se)
genes_dmr_df <- res_match$dmr_genes
genes_dnam_df <- res_match$dnam_genes

# evaluate associations
assoc_dmr_df <- calculate_association(dmr_se, rna_se, genes_dmr_df, adjust_braak = TRUE)
assoc_dnam_df <- calculate_association(dnam_se, rna_se, genes_dnam_df, adjust_braak = TRUE)

# annotate
assoc_dmr_df <- annotate_dmr(assoc_dmr_df, dmr_ref, rna_se) %>%
  dplyr::arrange(.data$pvalue)
assoc_dnam_df <- annotate_dnam(assoc_dnam_df, dnam_ref, rna_se) %>%
  dplyr::arrange(.data$pvalue)
```

### save

```{r save_all, eval = TRUE}
write.csv(
  assoc_dmr_df,
  file = file.path(result_dir, "DNAm_RNA", "All_DMR_to_RNA_Association.csv"),
  row.names = FALSE
)
write.csv(
  assoc_dnam_df,
  file = file.path(result_dir, "DNAm_RNA", "All_CpG_to_RNA_Association.csv"),
  row.names = FALSE
)
```

## Normal Data

### load data

```{r load_normal}
dnam_se <- readRDS(file = file.path(rosmap_dir, "rna_dnam", "dnam_normal_data.RDS"))
dmr_se <- readRDS(file = file.path(rosmap_dir, "rna_dnam", "dmr_norm_data.RDS"))
rna_se <- readRDS(file = file.path(rosmap_dir, "rna_dnam", "rnaseq_norm_data.RDS"))
```

### get associations

```{r calc_assocation_normal, eval = TRUE}
# match features
res_match <- match_features(dnam_se, dmr_se, rna_se)
genes_dmr_df <- res_match$dmr_genes
genes_dnam_df <- res_match$dnam_genes

# evaluate associations
assoc_dmr_df <- calculate_association(dmr_se, rna_se, genes_dmr_df, adjust_braak = TRUE)
assoc_dnam_df <- calculate_association(dnam_se, rna_se, genes_dnam_df, adjust_braak = TRUE)

# annotate
assoc_dmr_df <- annotate_dmr(assoc_dmr_df, dmr_ref, rna_se) %>%
  dplyr::arrange(.data$pvalue)
assoc_dnam_df <- annotate_dnam(assoc_dnam_df, dnam_ref, rna_se) %>%
  dplyr::arrange(.data$pvalue)
```

### save

```{r save_normal, eval = TRUE}
write.csv(
  assoc_dmr_df,
  file = file.path(result_dir, "DNAm_RNA", "Normal_DMR_to_RNA_Association.csv"),
  row.names = FALSE
)
write.csv(
  assoc_dnam_df,
  file = file.path(result_dir, "DNAm_RNA", "Normal_CpG_to_RNA_Association.csv"),
  row.names = FALSE
)
```

## AD Data

### load data

```{r load_ad}
dnam_se <- readRDS(file = file.path(rosmap_dir, "rna_dnam", "dnam_ad_data.RDS"))
dmr_se <- readRDS(file = file.path(rosmap_dir, "rna_dnam", "dmr_ad_data.RDS"))
rna_se <- readRDS(file = file.path(rosmap_dir, "rna_dnam", "rnaseq_ad_data.RDS"))
```

### get associations

```{r calc_assocation_ad, eval = TRUE}
# match features
res_match <- match_features(dnam_se, dmr_se, rna_se)
genes_dmr_df <- res_match$dmr_genes
genes_dnam_df <- res_match$dnam_genes

# evaluate associations
assoc_dmr_df <- calculate_association(dmr_se, rna_se, genes_dmr_df, adjust_braak = TRUE)
assoc_dnam_df <- calculate_association(dnam_se, rna_se, genes_dnam_df, adjust_braak = TRUE)

# annotate
assoc_dmr_df <- annotate_dmr(assoc_dmr_df, dmr_ref, rna_se) %>%
  dplyr::arrange(.data$pvalue)
assoc_dnam_df <- annotate_dnam(assoc_dnam_df, dnam_ref, rna_se) %>%
  dplyr::arrange(.data$pvalue)
```

### save

```{r save_ad, eval = TRUE}
write.csv(
  assoc_dmr_df,
  file = file.path(result_dir, "DNAm_RNA", "AD_DMR_to_RNA_Association.csv"),
  row.names = FALSE
)
write.csv(
  assoc_dnam_df,
  file = file.path(result_dir, "DNAm_RNA", "AD_CpG_to_RNA_Association.csv"),
  row.names = FALSE
)
```

# Display Result Summary Statistics

## Reload Results

```{r reload}
dmr_df <- read.csv(file.path(result_dir, "DNAm_RNA", "All_DMR_to_RNA_Association.csv"))
dnam_df <- read.csv(file.path(result_dir, "DNAm_RNA", "All_CpG_to_RNA_Association.csv"))

dmr_norm_df <- read.csv(file.path(result_dir, "DNAm_RNA", "Normal_DMR_to_RNA_Association.csv"))
dnam_norm_df <- read.csv(file.path(result_dir, "DNAm_RNA", "Normal_CpG_to_RNA_Association.csv"))

dmr_ad_df <- read.csv(file.path(result_dir, "DNAm_RNA", "AD_DMR_to_RNA_Association.csv"))
dnam_ad_df <- read.csv(file.path(result_dir, "DNAm_RNA", "AD_CpG_to_RNA_Association.csv"))
```

## Display Statistics

```{r result_stats}
data.frame(
  probe = c("CpG", "CpG", "CpG", "DMR", "DMR", "DMR"),
  dataset = c("All", "Normal", "AD", "All", "Normal", "AD"),
  correct_braak = c("Yes", "Yes", "Yes", "Yes", "Yes", "Yes"),
  comparisons = c(
    nrow(dnam_df),
    nrow(dnam_norm_df),
    nrow(dnam_ad_df),
    nrow(dmr_df),
    nrow(dmr_norm_df),
    nrow(dmr_ad_df)
  ),
  significant_fdr = c(
    sum(dnam_df$fdr <= 0.05),
    sum(dnam_norm_df$fdr <= 0.05),
    sum(dnam_ad_df$fdr <= 0.05),
    sum(dmr_df$fdr <= 0.05),
    sum(dmr_norm_df$fdr <= 0.05),
    sum(dmr_ad_df$fdr <= 0.05)
  ),
  significant_pvalue = c(
    sum(dnam_df$pvalue <= 0.05),
    sum(dnam_norm_df$pvalue <= 0.05),
    sum(dnam_ad_df$pvalue <= 0.05),
    sum(dmr_df$pvalue <= 0.05),
    sum(dmr_norm_df$pvalue <= 0.05),
    sum(dmr_ad_df$pvalue <= 0.05)
  )
) %>%
  dplyr::mutate(
    significant_expected = .data$comparisons / 20
  )
```

## Adjust Results to single CpGs

```{r result_adjust}
adjust_dnam <- function(dnam_df) {
  dnam_df %>%
    dplyr::arrange(.data$pvalue) %>%
    dplyr::group_by(.data$CpG) %>%
    dplyr::filter(dplyr::row_number() == 1) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(fdr = p.adjust(.data$pvalue, method = "fdr"))
}

dnam_df <- adjust_dnam(dnam_df)
dnam_norm_df <- adjust_dnam(dnam_norm_df)
dnam_ad_df <- adjust_dnam(dnam_ad_df)
```

## Display Adjusted Statistics

```{r result_adjust_stats}
data.frame(
  probe = c("CpG", "CpG", "CpG", "DMR", "DMR", "DMR"),
  dataset = c("All", "Normal", "AD", "All", "Normal", "AD"),
  correct_braak = c("Yes", "Yes", "Yes", "Yes", "Yes", "Yes"),
  comparisons = c(
    nrow(dnam_df),
    nrow(dnam_norm_df),
    nrow(dnam_ad_df),
    nrow(dmr_df),
    nrow(dmr_norm_df),
    nrow(dmr_ad_df)
  ),
  significant_fdr = c(
    sum(dnam_df$fdr <= 0.05),
    sum(dnam_norm_df$fdr <= 0.05),
    sum(dnam_ad_df$fdr <= 0.05),
    sum(dmr_df$fdr <= 0.05),
    sum(dmr_norm_df$fdr <= 0.05),
    sum(dmr_ad_df$fdr <= 0.05)
  ),
  significant_pvalue = c(
    sum(dnam_df$pvalue <= 0.05),
    sum(dnam_norm_df$pvalue <= 0.05),
    sum(dnam_ad_df$pvalue <= 0.05),
    sum(dmr_df$pvalue <= 0.05),
    sum(dmr_norm_df$pvalue <= 0.05),
    sum(dmr_ad_df$pvalue <= 0.05)
  )
) %>%
  dplyr::mutate(
    significant_expected = .data$comparisons / 20
  )
```

## Save Adjusted

```{r result_save}
write.csv(
  dnam_df,
  file = file.path(result_dir, "OLD", "CpG_to_RNA_Association_Adjusted.csv"),
  row.names = FALSE
)

write.csv(
  dnam_norm_df,
  file = file.path(result_dir, "OLD", "Normal_CpG_to_RNA_Association_Adjusted.csv"),
  row.names = FALSE
)

write.csv(
  dnam_ad_df,
  file = file.path(result_dir, "OLD", "AD_CpG_to_RNA_Association_Adjusted.csv"),
  row.names = FALSE
)
```

# Session Information

<details>
  <summary>**Session Info**</summary>
```{r session}
sessionInfo()
```
