---
title: "Create Miami Plot comparing Aging vs AD significant probes"
subtitle: "We correlate probes and genes on opposite strands"
author:
  - David Lukacsovich^[University of Miami]
  - Lily Wang^[University of Miami]
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  rmarkdown::html_document:
    highlight: breezedark
    theme: lumen
    toc: true
    number_sections: true
    df_print: paged
    code_download: false
    toc_float:
      collapsed: yes
    toc_depth: 3
editor_options:
  chunk_output_type: inline    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Set Up

## Load Libraries

```{r library}
suppressMessages({
  library(dplyr)
  library(miamiplot)
  library(ggplot2)
  library(readxl)
  library(IlluminaHumanMethylationEPICanno.ilm10b2.hg19)
})

base_dir <- "~/TBL Dropbox/David Lukacsovich"
analysis_dir <- file.path(base_dir, "AD-Aging-brain-sample-analysis")
result_dir <- file.path(analysis_dir, "analysis-results")
code_dir <- file.path(analysis_dir, "code")
dataset_dir <- file.path(code_dir, "DATASETS")
rosmap_dir <- file.path(dataset_dir, "Rosmap")
reference_dir <- file.path(code_dir, "Reference_Files")
zhang_dir <- file.path(
  base_dir,
  "coMethDMR_metaAnalysis/code_validation/Meta_analysis_code",
  "NatComm_revision/DATASETS"
)

source(file.path(code_dir, "R/ggmiami2.R"), local = TRUE)
```

## Load Data

```{r load}
ad_df <- read.csv(
  file.path(zhang_dir, "meta_analysis_single_cpg_df.csv")
) %>%
  dplyr::select(
    "cpg",
    chr = "seqnames",
    pos = "start",
    pval = "pVal.final",
    "fdr"
  ) %>%
  dplyr::filter(!is.na(.data$pval)) %>%
  dplyr::filter(!(.data$chr %in% c("chrM", "chrX", "chrY")))

annEPIC <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b2.hg19) %>%
  as.data.frame() %>%
    tibble::rownames_to_column(var = "cpg") %>%
    dplyr::select("cpg", gene = "UCSC_RefGene_Name")

ad_df <- ad_df %>%
  dplyr::left_join(annEPIC, by = "cpg")

aging_df <- read.csv(file = file.path(
  analysis_dir,
  "analysis-results",
  "meta_analysis",
  "meta_analysis_single_cpg_bacon.csv"
  )
) %>%
  dplyr::filter(.data$direction_bacon %in% c("--", "++", "-+", "+-"))
```

# Prepare Data

## Filter to Relevant Columns

```{r filter}
ad_df <- ad_df %>%
  dplyr::select(
    "cpg",
    "chr",
    "pos",
    "gene",
    "pval",
    "fdr"
  ) %>%
  dplyr::mutate(
    label = paste0(.data$chr, "_", .data$pos)
  )
aging_df <- aging_df %>%
  dplyr::select(
    cpg = "probe",
    chr = "seqnames",
    pos = "start",
    gene = "UCSC_RefGene_Name",
    pval = "pvalue_final_bacon",
    fdr = "fdr_bacon"
  ) %>%
  dplyr::mutate(
    label = paste0(.data$chr, "_", .data$pos)
  )
```

## Adjust Chromosomal Ranges

Miami-plot doesn't automatically align 2 datasets. If a chromosome is missing from one dataset, or the extreme positions of two chromosomes aren't the same, chromsomes will shift compared to each other. As such, we need to manually add in filler points - probes with no attached genes, and with p-values of 1 - at the observed extrema of each chromosome, to force both plots on teh same range

```{r adjust_range}
pmax <- mean(max(ad_df$pval), max(aging_df$pval))

chr_df <- rbind(
  ad_df %>% dplyr::select("chr", "pos"),
  aging_df %>% dplyr::select("chr", "pos")
) %>%
  dplyr::group_by(.data$chr) %>%
  dplyr::summarise(start = min(.data$pos), end = max(.data$pos)) %>%
  tidyr::pivot_longer(cols = c("start", "end"), names_to = "extrema", values_to = "pos") %>%
  dplyr::mutate(
    cpg = "",
    gene = "",
    pval = pmax,
    fdr = 1,
    label = paste0(.data$chr, "_", .data$pos)
  ) %>%
  dplyr::select(
    "cpg", "chr", "pos", "gene", "pval", "fdr", "label"
  )

ad_df <- rbind(
  ad_df,
  chr_df %>%
    dplyr::filter(!(.data$label %in% ad_df$label))
) %>%
  dplyr::mutate(label = "AD")

aging_df <- rbind(
  aging_df,
  chr_df %>%
    dplyr::filter(!(.data$label %in% aging_df$label))
) %>%
  dplyr::mutate(label = "Aging")
```

## Fix Label Genes

```{r fix_genes}
fix_gene <- function(gene) {
  gene[gene == "NA"] <- ""
  gene <- sapply(strsplit(gene, ";"), `[`, 1)
  gene[is.na(gene)] <- ""
  gene %>%
    stringr::str_replace("FLJ45983", "GATA3")
}

ad_df <- ad_df %>%
  dplyr::mutate(gene = fix_gene(.data$gene))

aging_df <- aging_df %>%
  dplyr::mutate(gene = fix_gene(.data$gene))
```


## Merge Two Dataframes and get Label Probes

```{r prepare_data}
miami_df <- rbind(ad_df, aging_df) %>%
  dplyr::mutate(seqnames = .data$chr) %>%
  dplyr::mutate(chr = as.integer(stringr::str_replace(.data$seqnames, "chr", "")))

plot_data <- miamiplot::prep_miami_data(
  data = miami_df,
  split_by = "label", 
  split_at = "Aging",
  p = "pval"
)

aging_labels <- plot_data$upper %>%
  dplyr::filter(nchar(.data$gene) > 0) %>%
  dplyr::arrange(dplyr::desc(.data$logged_p)) %>%
  dplyr::select("rel_pos", "logged_p", label = "gene") %>%
  head(n = 10)

ad_labels <- plot_data$lower %>%
  dplyr::filter(nchar(.data$gene) > 0) %>%
  dplyr::arrange(dplyr::desc(.data$logged_p)) %>%
  dplyr::select("rel_pos", "logged_p", label = "gene") %>%
  head(n = 10)

aging_unique_labels <- plot_data$upper %>%
  dplyr::filter(nchar(.data$gene) > 0) %>%
  dplyr::arrange(dplyr::desc(.data$logged_p)) %>%
  dplyr::group_by(.data$gene) %>%
  dplyr::slice(1) %>%
  dplyr::ungroup() %>%
  dplyr::arrange(dplyr::desc(.data$logged_p)) %>%
  dplyr::select("rel_pos", "logged_p", label = "gene") %>%
  head(n = 10)

ad_unique_labels <- plot_data$lower %>%
  dplyr::filter(nchar(.data$gene) > 0) %>%
  dplyr::arrange(dplyr::desc(.data$logged_p)) %>%
  dplyr::group_by(.data$gene) %>%
  dplyr::slice(1) %>%
  dplyr::ungroup() %>%
  dplyr::arrange(dplyr::desc(.data$logged_p)) %>%
  dplyr::select("rel_pos", "logged_p", label = "gene") %>%
  head(n = 10)
```

# Create Miami Plots

## Disallow Gene Repeats

```{r create_miami}
thresh_upper = aging_df %>%
  dplyr::filter(.data$fdr < 0.05) %>%
  dplyr::pull("pval") %>%
  max()

thresh_lower = ad_df %>%
  dplyr::filter(.data$fdr < 0.05) %>%
  dplyr::pull("pval") %>%
  max()

ggmiami2(
  data = miami_df, 
  split_by = "label",
  split_at = "Aging",
  p = "pval", 
  upper_ylab = "Aging-associated CpGs",
  lower_ylab = "AD-associated CpGs",
  upper_labels_df = aging_unique_labels,
  lower_labels_df = ad_unique_labels,
  suggestive_line_upper = thresh_upper,
  suggestive_line_bottom = thresh_lower,
  suggestive_line_color = "red",
  genome_line = NULL
)
```

## Save Plots

```{r}
pdf(
  file.path(
    result_dir, "comparison_aging_and_ad", "Aging vs AD Miami Plot.pdf"
  ),
  width = 8,
  height = 6
)
ggmiami2(
  data = miami_df, 
  split_by = "label",
  split_at = "Aging",
  p = "pval", 
  upper_ylab = "Aging-associated CpGs",
  lower_ylab = "AD-associated CpGs",
  upper_labels_df = aging_unique_labels,
  lower_labels_df = ad_unique_labels,
  suggestive_line_upper = thresh_upper,
  suggestive_line_bottom = thresh_lower,
  suggestive_line_color = "red",
  genome_line = NULL
)
dev.off()
```

# Session Information

<details>
  <summary>**Session Info**</summary>
```{r session}
sessionInfo()
```
</details>
